subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,792 @@
+{
+  GstRTPBasePayload *rtpbasepay = GST_RTP_BASE_PAYLOAD (rtph266pay);
+  GHashTableIter iter;
+  GstClockTime pts;
+  Nalu *nalu;
+
+  g_hash_table_iter_init (&iter, rtph266pay->xps_id_nalu_map);
+  while (g_hash_table_iter_next (&iter, NULL, (gpointer) & nalu)) {
+    g_queue_push_head (&rtph266pay->nalus, nalu_copy (nalu));
+
+    pts = GST_BUFFER_PTS (nalu->nalu_buf);
+    if (GST_CLOCK_TIME_IS_VALID (pts)) {
+      rtph266pay->last_xps_sent_ts =
+          gst_segment_to_running_time (&rtpbasepay->segment, GST_FORMAT_TIME,
+          pts);
+    }
+  }
+}
+
+static gboolean
+gst_rtp_h266_pay_has_to_send_xps (GstRtpH266Pay * rtph266pay, const Nalu * nalu)
+{
+  // TODO
+  // if nalu_type is not one of the allowed (is a non-VCL maybe?)
+  //  return FALSE;
+  //
+  // if (rtph266pay->last_xps_sent_ts < 0) && NALU_IS_IDR (nalu))
+  //  return TRUE;
+  //
+  // GstClockTime running_time =
+  //   gst_segment_to_running_time (&rtpbasepay->segment,
+  //       GST_FORMAT_TIME, pts);
+  //
+  // if (rtph265pay->last_xps_sent_ts + (rtph265pay->config_interval * GST_SECOND) >= running_time)
+  //   return TRUE;
+
+  return FALSE;
+}
+
+static void
+gst_rtp_h266_pay_set_au_boundaries (GstRtpH266Pay * rtph266pay)
+{
+  GList *head = g_queue_peek_head_link (&rtph266pay->nalus);
+  Nalu *last_nalu = g_queue_peek_tail (&rtph266pay->nalus);
+
+  for (GList * l = head; l != NULL; l = l->next) {
+    Nalu *nalu = l->data;
+    Nalu *prev_nalu = l->prev ? l->prev->data : NULL;
+    // TODO gboolean discont = GST_BUFFER_IS_DISCONT (nalu->nalu_buf);
+    GstClockTime prev_pts = GST_CLOCK_TIME_NONE;
+    GstClockTime prev_dts = GST_CLOCK_TIME_NONE;
+    GstClockTime pts = GST_BUFFER_PTS (nalu->nalu_buf);
+    GstClockTime dts = GST_BUFFER_DTS (nalu->nalu_buf);
+    if (prev_nalu) {
+      prev_pts = GST_BUFFER_PTS (prev_nalu->nalu_buf);
+      prev_dts = GST_BUFFER_DTS (prev_nalu->nalu_buf);
+    }
+    gboolean new_ts = (prev_pts != pts) || (prev_dts != dts);
+
+    nalu->au_start = new_ts /*|| discont */ ;
+    if (prev_nalu && nalu->au_start) {
+      prev_nalu->au_end = TRUE;
+      GST_DEBUG_OBJECT (rtph266pay, "AU start found -> previous AU finished");
+    }
+  }
+
+  // In some cases, we already know where is the end of an AU. Mark it now to
+  // avoid waiting for the next NALU.
+  if (last_nalu) {
+    gboolean au_alignment = rtph266pay->alignment == ALIGNMENT_AU;
+    gboolean marker =
+        GST_BUFFER_FLAG_IS_SET (last_nalu->nalu_buf, GST_BUFFER_FLAG_MARKER);
+    last_nalu->au_end = last_nalu->au_end || au_alignment || marker;
+  }
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_push_pending_data (GstRtpH266Pay * rtph266pay, gboolean eos)
+{
+  GQueue *nalus = &rtph266pay->nalus;
+  GstFlowReturn ret = GST_FLOW_OK;
+  Nalu *nalu;
+
+  // TODO: handle AP
+  // TODO: send XPS if needed
+
+  // TODO: Maybe it's worth to avoid iterating the NALU list twice, maybe not
+  gst_rtp_h266_pay_set_au_boundaries (rtph266pay);
```

Comment:
It should also work :crossed_fingers:. But we can't test it easily right now, as our encoder only supports `alignment=au` :/

Reply:
Then I would remove everything related to nal in the code, starting by removing `nal` alignment from the static pad template.

Reply:
That's true. Our encoder only supports `alignment=au`. We can only test with `alignment=au`. But our `rtph266pay` "supports" `alignment={nal, au}`.

This is meant to be a public contribution; the `gstrtph266pay` is not tightly bonded to our encoder. I think it's better to leave the `nal` support. Maybe someone else will test it in the future. And it might need to be fixed, or it could work :man_shrugging:.

Reply:
Only if tested, otherwise disable, or doing optional with a compilation macro, for example. IMO.

Reply:
I understand your point. However, my way of thinking about this is a bit different: Our code always contains untested logic, and we don't have 100% test coverage; not automatic, not manual. This is just another untested path.

I can't see any advantages in removing this option.